packages/node_modules/@ciscospark/spark-core/src/interceptors/rate-limit.js

```diff
@@ -0,0 +1,130 @@
+/*!
+ * Copyright (c) 2015-2017 Cisco Systems, Inc. See LICENSE file.
+ */
+
+import {Interceptor} from '@ciscospark/http-core';
+
+// contains the system time in milliseconds at which the retry after associated with a 429 expires
+// mapped by the API name, e.g.: idbroker.webex.com/doStuff would be mapped as 'doStuff'
+const rateLimitExpiryTime = new WeakMap();
+
+// extracts the common identity API being called
+const idBrokerRegex = /.*(idbroker|identity).webex.com\/([^\/]+)/;
+
+/**
+ * @class
+ */
+export default class RateLimitInterceptor extends Interceptor {
+
+  /**
+   * @returns {RateLimitInterceptor}
+   */
+  static create() {
+    return new RateLimitInterceptor({spark: this});
+  }
+
+  /**
+  * constructor
+  * @param {mixed} args
+  * @returns {Exception}
+  */
+  constructor(...args) {
+    super(...args);
+    rateLimitExpiryTime.set(this, new Map());
+  }
+    /**
+   * @see {@link Interceptor#onRequest}
+   * @param {Object} options
+   * @returns {Object}
+   */
+  onRequest(options) {
+    if (this.isRateLimited(options.uri)) {
+      return Promise.reject(
+        new Error(`API rate limited ${options.uri}`)
+      );
+    }
+    return Promise.resolve(options);
+  }
+  /**
+   * @see {@link Interceptor#onResponseError}
+   * @param {Object} options
+   * @param {Error} reason
+   * @returns {Object}
+   */
+  onResponseError(options, reason) {
+    if (reason.statusCode === 429 && (options.uri.includes(`idbroker`) || options.uri.includes(`identity`))) {
+      // set the retry after in the map, setting to milliseconds
+      this.setRateLimitExpiry(options.uri, this.extractRetryAfterTime(options) * 1000);
+    }
+    return Promise.reject(reason);
+  }
+
+  /**
+   * @param {object} options associated with the request
+   * @returns {number} retry after time in seconds
+   */
+  extractRetryAfterTime(options) {
+    const retryAfter = options.headers[`retry-after`] || null;
+    if (retryAfter === null || retryAfter <= 0) {
+      return 60;
+    }
+    else if (retryAfter > 3600) {
+      return 3600;
+    }
+    return retryAfter;
+  }
+
+  /**
+   * Sets the system time at which the rate limiting
+   * will expire in the rateLimitExpiryTime map.
+   * Assumes retryAfter is in milliseconds
+   * @param {string} uri API issuing the rate limiting
+   * @param {number} retryAfter milliseconds until rate limiting expires
+   * @returns {bool} true is value was successfully set
+   */
+  setRateLimitExpiry(uri, retryAfter) {
+    const apiName = this.getApiName(uri);
+    const currTimeMilli = new Date().getTime();
+    const expiry = currTimeMilli + retryAfter;
+    const dict = rateLimitExpiryTime.get(this);
+    return apiName && dict.set(apiName, expiry);
```

Comment:
Wouldn't the `apiName` check make more sense around line 87 to avoid potentially doing needless work if `apiName` is falsy?

Reply:
Changed.